File: app/src/main/java/com/flomobility/anx/hermes/common/Rate.kt
package com.flomobility.anx.hermes.common

data class Rate(val hz: Int) {

    fun toMicros(): Long {
        return (1_000_000L / hz)
    }

    fun toMillis(): Long {
        return (1_000L / hz)
    }

    private var _start = System.currentTimeMillis()
    private var _expected_cycle_time = 1000 / hz

    private var actual_cycle_time = 0L

    fun sleep() {
/*        val sleepDuration = (1000 / hz) - (System.currentTimeMillis() - time)
        if (sleepDuration >= 1) {
            Thread.sleep(sleepDuration)
        }
        time = System.currentTimeMillis()*/
        var end = _start + _expected_cycle_time
        val actualEnd = System.currentTimeMillis()

        if (actualEnd < _start) {
            end = actualEnd + _expected_cycle_time
        }

        val sleepDuration = end - actualEnd
        actual_cycle_time = actualEnd - _start

        _start = end
        if(sleepDuration <= 0) {
            if (actualEnd > end + _expected_cycle_time)
            {
                _start = actualEnd
            }
            return
        }
        Thread.sleep(sleepDuration)
    }
}

File: common/src/main/cpp/src/rate.cpp
//
// Created by Clay-Flo on 28/03/23.
//

#include "rate.h"

Rate::Rate(uint16_t hz)
        : start_(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()),
          expected_cycle_time_(1000 / hz),
          actual_cycle_time_(0L) {}

uint64_t Rate::get_stamp_in_ms() {
    auto stamp = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
    return stamp;
}

void Rate::sleep() {
    auto end = start_ + expected_cycle_time_;
    auto actual_end = get_stamp_in_ms();

    if (actual_end < start_) {
        end = actual_end + expected_cycle_time_;
    }

    auto sleep_duration = end - actual_end;
    actual_cycle_time_ = actual_end - start_;

    start_ = end;

    if (sleep_duration <= 0L) {
        if(actual_end > end + expected_cycle_time_) {
            start_ = actual_end;
        }
        return;
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(sleep_duration));
}
